lib/ansible/modules/cloud/oneandone/oneandone.py

```diff
@@ -0,0 +1,793 @@
+    """
+    Allow the incremental count in the description when defined with the
+    string formatting (%) operator. Otherwise, repeat the same description.
+    """
+    if '%' in description:
+        return [
+            description % i
+            for i in xrange(1, count + 1)
+        ]
+    else:
+        return [description] * count
+
+
+def _validate_custom_hardware_params(module):
+    for param in ('vcore',
+                  'cores_per_processor',
+                  'ram',
+                  'hdds'):
+        if not module.params.get(param):
+            return False
+    return True
+
+
+def _validate_hardware_params(module):
+    if (bool(module.params.get('fixed_instance_size')) ^
+       bool(_validate_custom_hardware_params(module))):
+        return True
+    return False
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            auth_token=dict(
+                type='str',
+                default=os.environ.get('ONEANDONE_AUTH_TOKEN')),
+            hostname=dict(type='str'),
+            description=dict(type='str'),
+            appliance=dict(type='str'),
+            fixed_instance_size=dict(type='str'),
+            vcore=dict(type='int'),
+            cores_per_processor=dict(type='int'),
+            ram=dict(type='float'),
+            hdds=dict(type='list'),
+            count=dict(type='int', default=1),
+            ssh_key=dict(type='raw', default=None),
+            auto_increment=dict(type='bool', default=True),
+            instance_ids=dict(type='list'),
+            datacenter=dict(
+                choices=DATACENTERS,
+                default='US'),
+            private_network=dict(type='str'),
+            firewall_policy=dict(type='str'),
+            load_balancer=dict(type='str'),
+            monitoring_policy=dict(type='str'),
+            wait=dict(type='bool', default=True),
+            wait_timeout=dict(type='int', default=600),
+            state=dict(type='str', default='present'),
+        )
+    )
+
+    if not HAS_ONEANDONE_SDK:
+        module.fail_json(msg='1and1 required for this module')
+
+    if not module.params.get('auth_token'):
+        module.fail_json(
+            msg='The "auth_token" parameter or ' +
+            'ONEANDONE_AUTH_TOKEN environment variable is required.')
+
+    oneandone_conn = oneandone.client.OneAndOneService(
+        api_token=module.params.get('auth_token'))
+
+    state = module.params.get('state')
+
+    if state == 'absent':
+        try:
+            (changed, machines) = remove_machine(module, oneandone_conn)
+        except Exception as e:
+            module.fail_json(msg=str(e))
+
+    elif state in ('running', 'stopped'):
+        try:
+            (changed, machines) = startstop_machine(module, oneandone_conn)
+        except Exception as e:
+            module.fail_json(msg=str(e))
+
+    elif state == 'present':
+        if not _validate_hardware_params(module):
+            module.fail_json(
```

Comment:
There are built in helper functions for option validation, see the following draft documentation:
https://github.com/gundalow/ansible/blob/docs-argspec/docs/docsite/rst/dev_guide/developing_modules_general.rst#main-and-ansiblemodule-argument-spec